apps/mason2/simulate_illumina.cpp

```diff
@@ -183,22 +186,13 @@ void _simulateSequence(TRead & read, TRng & rng, TFrag const & frag,
             continue;
         }
 
+        // NOTE: We can only insert CGAT, but we can have a polymorphism to N.
+        dist.param(param_t{0u, static_cast<unsigned short>(cigar[i].operation == 'I' ? 3u : 4u)});
```

Comment:
```suggestion
        dna_dist.param(dna_dist_param_t{0u, static_cast<dna_dist_int_t>(cigar[i].operation == 'I' ? 3u : 4u)});
```

Reply:
I think, this note is "wrong". This note shouldn't exists and an 'N' should never be drawn. It feels more like a documentation of a bug.
I really think that for substitution it was supposed to be something like: `x = (x + rand(1, 3)) % 4` (pseudo code).

In general this creates a huge doubt in whether mason actually represents any actual illumina sequence reader output. I can't imagine that this reflects how any of this works in the real world.

Reply:
Those are the four CIGARS simulated:
```
M = match     (consumes reference and read)
I = insertion (consumes read)
D = deletion  (consumes reference)
X = mismatch  (consumes reference and read)
```

It's a bit weird because `M` means alignment match, i.e. the sequence can be either a match or mismatch.
`=` would be a sequence match, which is how `M` is handled in the simulation.
`X` is a sequence mismatch.

M and D are already handled.

Seems like the intent is that an insertion cannot be `N` because it would end up just being an alignment match `M`.
But for a sequence mismatch, it could also be `N`.

Reply:
In general, Illumina reads can have Ns if the base call could not resolve the exact nucleotides. The question is, however, whether that follows the distribution given here, which is highly doubtable. So yes, Ns can be drawn, but what is the real distribution of this? This may also be more likely at the ends of a read or only for small fraction of reads that are close to the edge of the flow cell, etc.. 
So I have two takes on this. Either Mason simulates reads as given by sequencer, or simulates reads after fastqc quality control, where reads containing Ns are discarded (as @SGSSGene suggests). I am in favour of the last one as well as most scenarios would rather build on the assumption that reads have been run through quality control. It gives us the option to later add a dedicated pre-quality control option to Mason where we then may use a more realistic distribution of such failed base calls.
But removing the Ns from simulation now might require a bump of the minor version.

Reply:
I didn't find any other read simulator that simulates `N`.
mason1 also simulates `N`, but it has an option for turning it off (default is on).

Illumina can produce `N`, but this is more likely at the beginning or end of sequencing.

> But removing the Ns from simulation now might require a bump of the minor version.

I removed `N`s for now, but this release would be a version bump anyway?

Reply:
> I think it is fine to do it within version bump for the regular 2.5. release. Just out of curiosity. You updating it within the SeqAn2 apps, but I think @SGSSGene started to pull Mason out into a separate application under the SeqAn organization: https://github.com/seqan/mason. Which one is now the most current one?

Such a good question. I have almost zero memory about that mason fork/separation. We only work on it on 16th of May and 17the of May 2023 and then never changed anything again? super weird xD

Reply:
Simon forked, I noticed and was curious. Then we did something for a day and that was it.

The idea was to put mason into a separate repository because it gets frequent questions and bug reports.
So it would be easier to release new versions if it was separate from the seqan library.

Either this is something we should tackle after the 2.5.0 release, i.e., remove mason from seqan and point to the new repo; or we delete the fork to avoid confusion.